keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1131 @@
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Under this proposal, Kubernetes API servers and clients will support the Concise
+Binary Object Representation (CBOR) data format. CBOR will be available to
+clients as an alternative to JSON for serializing resources in request and
+response bodies. It will supersede JSON in apiextensions-apiserver for storage
+serialization of custom resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Reduce CPU time and heap churn of encode and decode for cases that can't use
+  Protobuf:
+  - custom resource storage
+  - custom resource serving
+  - dynamic clients
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Replace existing usage of Protobuf.
+- Substantially reduce the size of encoded objects (a modest size reduction is
+  anticipated).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Format
+
+The output of the CBOR encoder is a single tagged data item as specified in
+“[Self-Described
+CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor)”,
+with no additional envelope. Self-described CBOR -- a tagged data item with tag
+number 55799 -- has the same semantics as the same data item with no tag, with
+the convenient property that its encoded form is always prefixed by
+`0xd9d9f7`. By design, this prefix is never found at the beginning of a JSON
+text and can be used as a “magic number” to distinguish the data format of a
+stored object at rest.
+
+To support decoding custom resources that have been stored as a mixture of JSON
+and CBOR, the CBOR serializer will implement RecognizingDecoder by checking for
+the prefix `0xd9d9f7`.
+
+Streaming responses (i.e. watches) will be serialized as CBOR Sequences. A [CBOR
+Sequence](https://www.rfc-editor.org/rfc/rfc8742.html) is a concatenation of
+zero or more CBOR data items, with no additional framing. This is effectively
```

Comment:
Is this consistent with what we're doing for protobufs? If not, can you elaborate why?

Reply:
Protos are length prefixed because the encoded form of a message is just a sequence of fields. The proposal here is similar to our JSON framer, which concatenates one object after another and splits frames on JSON object boundaries (`{...}`). Our JSON framing scheme isn't based on a particular standard (at least according to https://github.com/kubernetes/design-proposals-archive/blob/acc25e14ca83dfda4f66d8cb1f1b491f26e78ffe/api-machinery/protobuf.md#streaming-wire-format).

My thinking is that, since there's been standardization work to define CBOR Sequences, it would be preferable to adopt that scheme vs. defining a CBOR framing scheme that is Kubernetes-specific. I do think that a length-prefixed framing is likely to have better performance because locating the start and end of a single data item requires either incremental parsing or two-pass parsing (the JSON framer finds frame boundaries by decoding into a `json.RawMessage` first).

Reply:
> that effectively makes each streamed item length-prefixed for efficient framing, without wrapping it in a custom envelope.

That is definitely worth considering, if we use it we have better opportunities for the future of taking advantage of framing, otherwise we will close that door...

Reply:
(as discussed out of band)

Watch events are encoded to a temporary buffer before being passed to the frame writer, and the frame writer can currently assume that each call to Write contains a complete frame. The Protobuf frame writer relies on this to determine its length prefix for free. One possible advantage of self-delimited framing / "no framing" is that the watch server can encode directly to the wire without copying every byte through a temporary buffer. There's no risk of sending partial objects, since a single object's encoding isn't valid without the final bytes (i.e. `}`). Streaming decode could also be possible -- the existing JSON frame reader adds an extra pass of the JSON unmarshaler over each frame to find the object boundaries that would be redundant with streaming decode.

Given all this, I'm leaning against specifying any extra framing or tagging, unless you think it won't be possible to eventually remove framer buffering for self-delimiting formats?

Reply:
Done, added the argument for unbuffered streaming.